### ISSUE TITLE
IOSXRDriver Timed out looking for SSH login password prompt

### ISSUE THREAD

**Issue by dorko20** [carlmontanari/scrapli#21]
**Describe the bug**
When using the IOSXRDriver with ssh-config and ssh-key type authentication  I got `TimeoutError: Timed out looking for SSH login password prompt`.   
Strange is that in debug I can see the scrapli did actually succeed to connect to my router but seems like it doesn't match the returned prompt and failed.  I can also see in my router's log that test user did connect succesfully.  
What is also interesting that when I try to connect to another IOS-XR router via username/password (no ssh-config, no ssh-key) then everything works ok.

**To Reproduce**
1. Your script  
```python
from scrapli.driver.core import IOSXRDriver
import logging

logging.basicConfig(level=logging.DEBUG)
logger = logging.getLogger("scrapli")

args = { 
    "host": "lab-test-r2n0-tt", 
    "auth_strict_key": False,
    "ssh_config_file": True,
    "timeout_socket": 15,
} 
conn = IOSXRDriver(**args)
conn.open()

INFO:driver:Opening connection to lab-test-r2n0-tt
INFO:transport:Attempting to authenticate to lab-test-r2n0-tt
INFO:transport:Attempting to open session with the following command: ['ssh', 'lab-test-r2n0-tt', '-p', '22', '-o', 'ConnectTimeout=15', '-o', 'ServerAliveInterval=10', '-o', 'StrictHostKeyChecking=no', '-o', 'UserKnownHostsFile=/dev/null', '-F', '/home/test/.ssh/config']
DEBUG:transport:Session to host lab-test-r2n0-tt spawned
DEBUG:transport:Attempting to authenticate. Read: b'\r\n\r\n'
DEBUG:transport:Attempting to authenticate. Read: b'RP/0/RP0/CPU0:lab-test-r2n0-tt#'
INFO:scrapli:Timed out looking for SSH login password prompt
INFO:scrapli:timeout_exit is True, closing transport
DEBUG:transport:Channel to host lab-test-r2n0-tt closed
CRITICAL:transport:Failed to open connection to host lab-test-r2n0-tt. Do you need to disable `auth_strict_key`?
---------------------------------------------------------------------------
TimeoutError                              Traceback (most recent call last)
<ipython-input-1-884c1bd1799b> in <module>
     12 } 
     13 conn = IOSXRDriver(**args)
---> 14 conn.open()

/usr/local/lib/python3.6/dist-packages/scrapli/driver/driver.py in open(self)
    575         """
    576         LOG.info(f"Opening connection to {self._initialization_args['host']}")
--> 577         self.transport.open()
    578         if self.on_open:
    579             self.on_open(self)

/usr/local/lib/python3.6/dist-packages/scrapli/transport/systemssh.py in open(self)
    315 
    316         # If public key auth fails or is not configured, open a pty session
--> 317         if not self._open_pty():
    318             msg = f"Authentication to host {self.host} failed"
    319             LOG.critical(msg)

/usr/local/lib/python3.6/dist-packages/scrapli/transport/systemssh.py in _open_pty(self, skip_auth)
    453         if skip_auth:
    454             return True
--> 455         self._pty_authenticate(pty_session=self.session)
    456         if not self._pty_isauthenticated(self.session):
    457             return False

/usr/local/lib/python3.6/dist-packages/scrapli/decorators.py in timeout_wrapper(channel_or_transport, *args, **kwargs)
     75                         session_lock.release()
     76                     close()
---> 77                 raise TimeoutError(message)
     78             return future.result()
     79 

TimeoutError: Timed out looking for SSH login password prompt

```
2. What you're connecting to (vendor, platform, version)
```sh
RP/0/RP0/CPU0:lab-test-r2n0-tt#show version 
Thu May 28 21:23:13.760 CEST
Cisco IOS XR Software, Version 6.6.3
Copyright (c) 2013-2019 by Cisco Systems, Inc.

Build Information:
 Built By     : hlo
 Built On     : Fri Dec 13 17:04:06 PST 2019
 Built Host   : iox-ucs-032
 Workspace    : /auto/srcarchive15/prod/6.6.3/asr9k-x64/ws
 Version      : 6.6.3
 Location     : /opt/cisco/XR/packages/
 Label        : 6.6.3-0

cisco ASR9K () processor
System uptime is 6 weeks 2 days 6 hours 

```
3. Anything else relevant
- not sure if relevant but in order to connect to my router I need to use a jumphost which is specified in my ssh config file as follow:  
```sh
host jumphost
  HostName jumphost
  User admin

Host lab-test-r2n0-tt
    Hostname 10.10.10.1
    ProxyCommand ssh -q -W %h:%p jumphost

# Default parameters
host *
    User cisco
    IdentityFile ~/.ssh/id_rsa
    IdentitiesOnly yes
    CheckHostIP no
    PasswordAuthentication yes
    ConnectTimeout 20
    StrictHostKeyChecking no
    ForwardAgent yes
```

**Expected behavior**
Scrapli should be able connect to the IOSXR device with the existing ssh config.

**Stack Trace**
Copy of your stack trace here, please format it properly using triple back ticks (top left key on US keyboards!)

**Screenshots**
If applicable, add screenshots to help explain your problem, but do note that formatted text is much preferred over
 screenshots!

**OS (please complete the following information):**
 - OS: `linux mint 19.3 `  
 - scrapli version: `2020.5.9`  
 - version of any optional extras (paramiko|ssh2-python|textfsm, etc.)

**Additional context**

**Comment by carlmontanari** [comment]
Hey @dorko20 thanks for the bug report! 

On quick look I think what is happening is we dont find a ssh key (even though it obviously exists) so scrapli tries to password auth and fails because of course it never finds the password prompt. 

I think my kinda messy custom implementation of ssh config parser is to blame for this -- I suspect (but need to verify) that we find the hostname in the ssh config but since the key is not associated to that host we dont check the `*` entry. 

This is a really good motivator for me to probably re-do a bit of the `SystemSSH` auth stuff and the ssh config parser, but that probably will take me a bit to get to....

In the short term I think you could set `auth_bypass` to True in your args dictionary, this would cause scrapli to just skip the whole authentication/verification process. Obviously not a great fix, but I believe it would get you going while I re-work the other bits. Do note that this is only available on the `system` transport option though.

Could you give that a try and let me know if that works, and if not some more logs would be fantastic!

Thanks!
Carl

**Comment by dorko20** [comment]
Hi Carl,
thanks a lot for you answer, I tried to set the `"auth_bypass": True` and now it is working:

```python
In [3]: from scrapli.driver.core import IOSXRDriver 
   ...: import logging 
   ...:  
   ...: logging.basicConfig(level=logging.DEBUG) 
   ...: logger = logging.getLogger("scrapli") 
   ...:  
   ...: args = {  
   ...:     "host": "lab-test-r2n0-tt",  
   ...:     "auth_strict_key": False, 
   ...:     "ssh_config_file": True, 
   ...:     "timeout_socket": 15, 
   ...:     "auth_bypass": True, 
   ...: }  
   ...: conn = IOSXRDriver(**args) 
   ...: conn.open()                           

INFO:driver:Opening connection to lab-test-r2n0-tt
INFO:transport:Attempting to authenticate to lab-test-r2n0-tt
INFO:transport:`auth_bypass` is True, bypassing authentication
INFO:transport:Attempting to open session with the following command: ['ssh', 'lab-test-r2n0-tt', '-p', '22', '-o', 'ConnectTimeout=15', '-o', 'ServerAliveInterval=10', '-o', 'StrictHostKeyChecking=no', '-o', 'UserKnownHostsFile=/dev/null', '-F', '/home/test/.ssh/config']
DEBUG:transport:Session to host lab-test-r2n0-tt spawned
INFO:driver:Attempting to acquire `privilege_exec` privilege level
DEBUG:channel:Write (sending return character): '\n'
DEBUG:channel:Read: b'\n'
DEBUG:channel:Read: b'\n\n'
DEBUG:channel:Read: b'RP/0/RP0/CPU0:lab-test-r2n0-tt#'
DEBUG:driver:Current privilege level could be `privilege_exec`
DEBUG:driver:Determined current privilege level is one of `['privilege_exec']`
INFO:driver:Acquired requested privilege level `privilege_exec`
INFO:channel:Attempting to send input: terminal length 0; strip_prompt: True
DEBUG:channel:Write: 'terminal length 0'
DEBUG:channel:Read: b'terminal length 0'
INFO:channel:Read: b'terminal length 0'
DEBUG:channel:Write (sending return character): '\n'
DEBUG:channel:Read: b'\n'
DEBUG:channel:Read: b'Fri May 29 07:56:53.539 CEST\n'
DEBUG:channel:Read: b'RP/0/RP0/CPU0:lab-test-r2n0-tt#'
INFO:channel:Read: b'\nFri May 29 07:56:53.539 CEST\nRP/0/RP0/CPU0:lab-test-r2n0-tt#'
INFO:channel:Attempting to send input: terminal width 512; strip_prompt: True
DEBUG:channel:Write: 'terminal width 512'
DEBUG:channel:Read: b'terminal width 512'
INFO:channel:Read: b'terminal width 512'
DEBUG:channel:Write (sending return character): '\n'
DEBUG:channel:Read: b'\n'
DEBUG:channel:Read: b'Fri May 29 07:56:53.888 CEST\n'
DEBUG:channel:Read: b'RP/0/RP0/CPU0:lab-test-r2n0-tt#'
INFO:channel:Read: b'\nFri May 29 07:56:53.888 CEST\nRP/0/RP0/CPU0:lab-test-r2n0-tt#'
INFO:driver:Connection to lab-test-r2n0-tt opened successfully
```

In the meantime I also tried to set the `"auth_private_key"` and `"auth_username"` attributes to the IOSXRDriver but with no luck, without the  `"auth_bypass": True` , it doens' work.
```python
In [6]: #conn.close() 
   ...: #del conn 
   ...: from scrapli.driver.core import IOSXRDriver 
   ...: import logging 
   ...:  
   ...: logging.basicConfig(level=logging.DEBUG) 
   ...: logger = logging.getLogger("scrapli") 
   ...:  
   ...: args = {  
   ...:     "host": "lab-test-r2n0-tt",  
   ...:     "auth_strict_key": False, 
   ...:     "ssh_config_file": True, 
   ...:     "timeout_socket": 15, 
   ...:     "auth_private_key": "/home/test/.ssh/id_rsa", 
   ...:     "auth_username": "cisco", 
   ...: }  
   ...: conn = IOSXRDriver(**args) 
   ...: conn.open()                                                                                                                                                                                  
INFO:driver:Opening connection to lab-test-r2n0-tt
INFO:transport:Attempting to authenticate to lab-test-r2n0-tt
INFO:transport:Attempting to open session with the following command: ['ssh', 'lab-test-r2n0-tt', '-p', '22', '-o', 'ConnectTimeout=15', '-o', 'ServerAliveInterval=10', '-i', '/home/test/.ssh/id_rsa', '-l', 'cisco', '-o', 'StrictHostKeyChecking=no', '-o', 'UserKnownHostsFile=/dev/null', '-F', '/home/test/.ssh/config', '-v', '-o', 'BatchMode=yes']
DEBUG:transport:Session to host lab-test-r2n0-tt spawned
INFO:scrapli:Timed out determining if session is authenticated
INFO:scrapli:timeout_exit is True, closing transport
DEBUG:transport:Channel to host lab-test-r2n0-tt closed
CRITICAL:transport:Failed to authenticate to host lab-test-r2n0-tt with private key `/home/test/.ssh/id_rsa`. Unable to continue authentication, missing username, password, or both.
---------------------------------------------------------------------------
ScrapliAuthenticationFailed               Traceback (most recent call last)
<ipython-input-6-539b1f0116ec> in <module>
     16 } 
     17 conn = IOSXRDriver(**args)
---> 18 conn.open()

/usr/local/lib/python3.6/dist-packages/scrapli/driver/driver.py in open(self)
    575         """
    576         LOG.info(f"Opening connection to {self._initialization_args['host']}")
--> 577         self.transport.open()
    578         if self.on_open:
    579             self.on_open(self)

/usr/local/lib/python3.6/dist-packages/scrapli/transport/systemssh.py in open(self)
    307                 )
    308                 LOG.critical(msg)
--> 309                 raise ScrapliAuthenticationFailed(msg)
    310             msg = (
    311                 f"Failed to authenticate to host {self.host} with private key "

ScrapliAuthenticationFailed: Failed to authenticate to host lab-test-r2n0-tt with private key `/home/test/.ssh/id_rsa`. Unable to continue authentication, missing username, password, or both.
```

**Comment by carlmontanari** [comment]
Thanks for the update! Glad its at least working with auth_bypass. I'll hopefully be able to test and recreate this this weekend or next weekend and if I can recreate it (guessing that should be no problem) will try to get it fixed and pushed out w/ the next pypi release.

I'll leave this open and update it when I have more info, thanks again for the detailed report!

**Comment by carlmontanari** [comment]
Hello again!

This should now be fixed in master/latest pypi version (2020.06.06).

The fix at least for the first portion of this issue was to add support for the ssh config file parser to "merge" attributes for a host. Given a config file like yours:

```
host jumphost
  HostName jumphost
  User admin

Host lab-test-r2n0-tt
    Hostname 10.10.10.1
    ProxyCommand ssh -q -W %h:%p jumphost

# Default parameters
host *
    User cisco
    IdentityFile ~/.ssh/id_rsa
    IdentitiesOnly yes
    CheckHostIP no
    PasswordAuthentication yes
    ConnectTimeout 20
    StrictHostKeyChecking no
    ForwardAgent yes
```

scrapli would previously not associate the `IdentityFile` from the `*` entry with the host `lab-test-r2n0-tt` -- this has now been updated to ensure that all subsequent config sections of the ssh config file are parsed/loaded up for a given host. 

Note that scrapli still does not "fully" parse the ssh config file -- in the case of system transport we really just care about finding a key for auth so that we can use the `pipes` authentication process of the system transport as this is more reliable than the default `pty` authentication process. 

In any case, the scenario you provided *should* be working with the currently available scrapli version, so please let me know if there are any remaining issues!

I'm still a bit unsure what caused things to fail when you explicitly passed the ssh key to use for auth though. If that is still an issue for whatever reason would you mind opening a new issue and we can figure it out there?

Thanks again for submitting this, I'll go ahead and close it for now, but if its still an issue feel free to reopen!

Carl
